### ISSUE TITLE
[GridSample] GridSample operation gives different outputs between onnx and tflite models

### ISSUE THREAD

**Issue by eugval** [PINTO0309/onnx2tf#274]
### Issue Type

Others

### onnx2tf version number

1.7.33

### onnx version number

1.12.0

### tensorflow version number

2.11.0

### Download URL for ONNX

N/A

### Parameter Replacement JSON

```json
N/A
```

### Description

There seems to be an issue with the behaviour of GridSample, which is not the same between the onnx model and the tensorflow model. This appears to be a bug?

To be clear the grid_sample converts, and the output array is the right shape, but the actual values in the output array are different between the tensorflow and pytorch versions of the operation. 

What is even more strange, is that the accuracy check with the `check_onnx_tf_outputs_elementwise_close_full` flag passes.

It would be great if this issue could be fixed, because it currently seems that pytorch models with the grid_sample operation will break when converted to tflite.

Here is the python script to reproduce:
pytorch version : 1.13.1
numpy version: 1.22.4
onnxruntime version: 1.13.1

This script creates a model with just the grid_sample operation. It then converts it to onnx, then to tflite, and compares the outputs for all three versions.

```python
import torch
import onnx2tf
import onnxruntime
import torch.nn.functional as F
import numpy as np
import tensorflow as tf

class Model(torch.nn.Module):
    def forward(self, input_tensor, grid_tensor):
        return F.grid_sample(
            input=input_tensor,
            grid=grid_tensor,
            padding_mode="zeros",
            mode="bilinear",
            align_corners=True,
        )


example_input_tensor = torch.randn(8,64,32,32)
example_grid_tensor = torch.rand((8, 40, 60, 2))

# Check PyTorch output
model = Model()
pytorch_output = model.forward(example_input_tensor, example_grid_tensor)

# Export to ONNX
torch.onnx.export(
    Model(),
    {"input_tensor": example_input_tensor, "grid_tensor": example_grid_tensor},
    "model.onnx",
    opset_version=16,
    input_names=["input_tensor", "grid_tensor"],
    output_names=["output_tensor"],
)

# ONNX works fine
session = onnxruntime.InferenceSession("model.onnx")
onnx_output = session.run(
    ["output_tensor"],
    {
        "input_tensor": example_input_tensor.numpy(),
        "grid_tensor": example_grid_tensor.numpy(),
    },
)


# ONNX2TF behaviour
onnx2tf.convert(
    input_onnx_file_path="model.onnx",
    output_folder_path="model.tf",
    non_verbose=False,
    keep_shape_absolutely_input_names=["grid_tensor"],
    check_onnx_tf_outputs_elementwise_close_full= True,
    check_onnx_tf_outputs_elementwise_close_atol= 1e-7
)


## Load tflite model and compare outputs
tflite_path = 'model.tf/model_float32.tflite'
interpreter = tf.lite.Interpreter(model_path = tflite_path)


interpreter.allocate_tensors()


input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()


input_shape1 = input_details[0]['shape']
input_data1 =example_input_tensor.permute(0,2,3,1).detach().numpy().astype(np.float32)
interpreter.set_tensor(input_details[0]['index'], input_data1)


input_shape2 = input_details[1]['shape']
input_data2 = example_grid_tensor.detach().numpy().astype(np.float32)
interpreter.set_tensor(input_details[1]['index'], input_data2)


interpreter.invoke()

output_data = interpreter.get_tensor(output_details[0]['index'])


print(f'Onnx outputs error vs pytorch: { np.mean(np.abs(onnx_output[0]- pytorch_output.detach().numpy()))}')
print(f"Tflite outputs error vs pytorch: {np.mean(np.abs(pytorch_output.permute(0,2,3,1).detach().numpy()-output_data))}")
print(f"Tflite outputs error vs onnx: {np.mean(np.abs(np.transpose(onnx_output[0], axes = (0,2,3,1))-output_data))}")


print('DONE')
```
Here is the output:
![image](https://user-images.githubusercontent.com/24697174/227980972-cd21d5c1-8509-4efb-8f7e-577d65fd4be9.png)

We can see that:
- The Onnx and pytorch outputs match.
- The tflite and pytorch outputs do not match
- The tflite and onnx outputs do not match

Thanks  for the help and the great repo!

**Comment by PINTO0309** [comment]
Yes. It is a bug. Please submit a pull request.

**Comment by PINTO0309** [comment]
This will work for `bilinear` `gridsample`, so you can try replacing it if you are in a hurry.
```python
 def bilinear_sample_noloop(image, grid):
      """
      :param image: sampling source of shape [N, C, H, W]
      :param grid: integer sampling pixel coordinates of shape [N, grid_H, grid_W, 2]
      :return: sampling result of shape [N, C, grid_H, grid_W]
      """
      Nt, C, H, W = image.shape
      grid_H = grid.shape[1]
      grid_W = grid.shape[2]
      xgrid, ygrid = grid.split([1, 1], dim=-1)
      mask = ((xgrid >= 0) & (ygrid >= 0) & (xgrid < W - 1) & (ygrid < H - 1)).float()
      x0 = torch.floor(xgrid)
      x1 = x0 + 1
      y0 = torch.floor(ygrid)
      y1 = y0 + 1
      wa = ((x1 - xgrid) * (y1 - ygrid)).permute(3, 0, 1, 2)
      wb = ((x1 - xgrid) * (ygrid - y0)).permute(3, 0, 1, 2)
      wc = ((xgrid - x0) * (y1 - ygrid)).permute(3, 0, 1, 2)
      wd = ((xgrid - x0) * (ygrid - y0)).permute(3, 0, 1, 2)
      x0 = (x0 * mask).view(Nt, grid_H, grid_W).long()
      y0 = (y0 * mask).view(Nt, grid_H, grid_W).long()
      x1 = (x1 * mask).view(Nt, grid_H, grid_W).long()
      y1 = (y1 * mask).view(Nt, grid_H, grid_W).long()
      ind = torch.arange(Nt, device=image.device) #torch.linspace(0, Nt - 1, Nt, device=image.device)
      ind = ind.view(Nt, 1).expand(-1, grid_H).view(Nt, grid_H, 1).expand(-1, -1, grid_W).long()
      image = image.permute(1, 0, 2, 3)
      output_tensor = (
          image[:, ind, y0, x0] * wa + image[:, ind, y1, x0] * wb + image[:, ind, y0, x1] * wc + image[:, ind, y1, x1] * wd
      ).permute(1, 0, 2, 3)
      output_tensor *= mask.permute(0, 3, 1, 2).expand(-1, C, -1, -1)
      image = image.permute(1, 0, 2, 3)
      return output_tensor, mask

  output, _ = bilinear_sample_noloop(x, vgrid)
```

**Comment by PINTO0309** [comment]
```
onnx=====================================
[[[[ 5.292332e-01  6.328660e-01  7.590018e-01 ...  2.709882e-01  7.027159e-01  2.210859e-01]
   [ 8.450945e-01  2.147626e-01  8.066105e-01 ...  7.200179e-01  4.355544e-01  4.737483e-01]
   [ 1.665662e-01  3.073401e-01  3.990825e-01 ...  8.667974e-01  6.447428e-01  1.089618e-01]

tflite===================================
[[[[ 5.292332e-01  6.328660e-01  7.590018e-01 ...  2.709882e-01  7.027159e-01  2.210859e-01]
   [ 8.450945e-01  2.147626e-01  8.066105e-01 ...  7.200179e-01  4.355544e-01  4.737483e-01]
   [ 1.665662e-01  3.073401e-01  3.990825e-01 ...  8.667974e-01  6.447428e-01  1.089618e-01]
```

**Comment by PINTO0309** [comment]
This is simply reimplemented in TensorFlow (Keras).
![image](https://user-images.githubusercontent.com/33194443/228901967-8ff57c5f-1dda-4925-9043-3ca31e110686.png)

**Comment by On-JungWoan** [comment]
@PINTO0309 
I think the replacement script is modified like this.

```python
    ...

    device = image.device

    ...

    # Clip coordinates to padded image size
    x0 = torch.where(x0 < 0, torch.tensor(0).to(device), x0)
    x0 = torch.where(x0 > padded_w - 1, torch.tensor(padded_w - 1).to(device), x0)
    x1 = torch.where(x1 < 0, torch.tensor(0).to(device), x1)
    x1 = torch.where(x1 > padded_w - 1, torch.tensor(padded_w - 1).to(device), x1)
    y0 = torch.where(y0 < 0, torch.tensor(0).to(device), y0)
    y0 = torch.where(y0 > padded_h - 1, torch.tensor(padded_h - 1).to(device), y0)
    y1 = torch.where(y1 < 0, torch.tensor(0).to(device), y1)
    y1 = torch.where(y1 > padded_h - 1, torch.tensor(padded_h - 1).to(device), y1)
```

**Comment by PINTO0309** [comment]
```
onnx2tf -i model.onnx -kat grid_tensor
```
```
Onnx outputs error vs pytorch: 2.698032375292314e-08
Tflite outputs error vs pytorch: 1.639568658617918e-08
Tflite outputs error vs onnx: 2.89812760456698e-08
DONE
```
Fixes: https://github.com/PINTO0309/onnx2tf/releases/tag/1.8.6

**Comment by Shakarim94** [comment]
@PINTO0309 really appreciate the update! Can you please share the readable code written in TF? Similar to the pytorch code you shared
